File: node-src/collectors/rest-diagnostics.ts
import type { HttpTransport } from "./types.js";
import {
  createRestHttpTransport,
  REST_HTTP_TIMEOUT_MS,
  type RestHttpTransportOptions,
} from "../io/rest-http-transport.js";

export interface RestHostProbeResult {
  readonly detail: string;
  readonly host: string;
  readonly ok: boolean;
}

export interface RestHostProbeOptions {
  readonly environment?: Readonly<Record<string, string | undefined>>;
  readonly includeCustomCa?: boolean;
  readonly timeoutMs?: number;
  /** Test seam; callers that supply a transport retain ownership of it. */
  readonly transport?: HttpTransport;
  /** Additional production-transport test seams. */
  readonly transportOptions?: Omit<
    RestHttpTransportOptions,
    "includeCustomCa" | "requestTimeoutMs"
  >;
}

function hostUrl(host: string): URL {
  if (
    typeof host !== "string"
    || host.length === 0
    || host.includes("/")
    || host.includes("@")
    || host.includes("?")
    || host.includes("#")
  ) {
    throw new TypeError("diagnostic host must be a hostname with an optional port");
  }
  let url: URL;
  try {
    url = new URL(`https://${host}/`);
  } catch {
    throw new TypeError("diagnostic host must be a hostname with an optional port");
  }
  if (url.hostname === "" || url.username !== "" || url.password !== "") {
    throw new TypeError("diagnostic host must be a hostname with an optional port");
  }
  return url;
}

/** Probe one collector host; any HTTP response proves DNS/TCP/TLS success. */
export async function probeRestHost(
  host: string,
  options: RestHostProbeOptions = {},
): Promise<RestHostProbeResult> {
  const url = hostUrl(host);
  const owned = options.transport === undefined;
  const transport = options.transport ?? await createRestHttpTransport(
    options.environment ?? process.env,
    {
      ...options.transportOptions,
      includeCustomCa: options.includeCustomCa !== false,
      requestTimeoutMs: options.timeoutMs ?? Math.min(15_000, REST_HTTP_TIMEOUT_MS),
    },
  );
  let result: RestHostProbeResult;
  try {
    const response = await transport.request({
      headers: { accept: "*/*" },
      method: "GET",
      timeoutMs: options.timeoutMs ?? Math.min(15_000, REST_HTTP_TIMEOUT_MS),
      url,
    });
    result = Object.freeze({
      detail: `HTTP ${response.status}`,
      host,
      ok: true,
    });
  } catch (error: unknown) {
    result = Object.freeze({
      detail: error instanceof Error ? error.message : "connection failed",
      host,
      ok: false,
    });
  } finally {
    if (owned) {
      try {
        await transport.close?.();
      } catch {
        // The connectivity result remains useful if best-effort cleanup fails.
      }
    }
  }
  return result;
}

/** Probe a deterministic host list without sharing cookies or connections. */
export async function probeRestHosts(
  hosts: readonly string[],
  options: RestHostProbeOptions = {},
): Promise<readonly RestHostProbeResult[]> {
  const results: RestHostProbeResult[] = [];
  for (const host of [...new Set(hosts)].sort()) {
    results.push(await probeRestHost(host, options));
  }
  return Object.freeze(results);
}

File: go/internal/collectors/rest_diagnostics.go
package collectors

import (
	"errors"
	"fmt"
	"net/url"
	"strings"

	"github.com/dvmrry/infrawright-dev/go/internal/canonjson"
)

// The fetch-diag surface validates hosts and probes connectivity through an
// injected transport. The CLI owns real-transport construction and lifetime;
// these helpers own deterministic host selection and the rule that any HTTP
// response proves DNS/TCP/TLS connectivity.

// defaultProbeTimeoutMs bounds an individual diagnostic request at 15 seconds.
const defaultProbeTimeoutMs = 15_000

// RestHostProbeResult describes one host connectivity probe.
type RestHostProbeResult struct {
	Detail string
	Host   string
	OK     bool
}

// RestHostProbeOptions configures a host probe. Transport is required;
// TimeoutMs == 0 uses defaultProbeTimeoutMs.
type RestHostProbeOptions struct {
	TimeoutMs int
	Transport HttpTransport
}

// hostURL validates a host and converts it to an HTTPS root URL.
func hostURL(host string) (*url.URL, error) {
	invalid := errors.New("diagnostic host must be a hostname with an optional port")
	if host == "" || strings.ContainsAny(host, "/@?#") {
		return nil, invalid
	}
	parsed, err := url.Parse("https://" + host + "/")
	if err != nil {
		return nil, invalid
	}
	if parsed.Hostname() == "" || parsed.User != nil {
		return nil, invalid
	}
	return parsed, nil
}

// ProbeRestHost probes one collector host; any HTTP response proves
// DNS/TCP/TLS success. ProbeRestHost never closes the caller-supplied
// transport.
func ProbeRestHost(host string, options RestHostProbeOptions) (RestHostProbeResult, error) {
	target, err := hostURL(host)
	if err != nil {
		return RestHostProbeResult{}, err
	}
	if options.Transport == nil {
		return RestHostProbeResult{}, errors.New("ProbeRestHost requires an injected transport")
	}
	timeoutMs := options.TimeoutMs
	if timeoutMs == 0 {
		timeoutMs = defaultProbeTimeoutMs
	}
	response, requestErr := options.Transport.Request(HTTPRequest{
		Method:    "GET",
		URL:       target,
		Headers:   map[string]string{"accept": "*/*"},
		TimeoutMs: timeoutMs,
	})
	if requestErr != nil {
		return RestHostProbeResult{Detail: requestErr.Error(), Host: host, OK: false}, nil
	}
	return RestHostProbeResult{Detail: fmt.Sprintf("HTTP %d", response.Status), Host: host, OK: true}, nil
}

// ProbeRestHosts probes a deterministic host list through the supplied
// transport.
func ProbeRestHosts(hosts []string, options RestHostProbeOptions) ([]RestHostProbeResult, error) {
	unique := make(map[string]struct{}, len(hosts))
	for _, host := range hosts {
		unique[host] = struct{}{}
	}
	names := make([]string, 0, len(unique))
	for host := range unique {
		names = append(names, host)
	}
	// Hostnames are ASCII, so the canonical string ordering is sufficient.
	names = canonjson.SortedStrings(names)
	results := make([]RestHostProbeResult, len(names))
	for i, host := range names {
		result, err := ProbeRestHost(host, options)
		if err != nil {
			return nil, err
		}
		results[i] = result
	}
	return results, nil
}
